pkg/hub/controllers/finalizercontroller/add_finalizer_controller.go

```diff
@@ -0,0 +1,171 @@
+	controller := &AddFinalizerController{
+		roleLister:        roleInformer.Lister(),
+		roleBindingLister: roleBindingInformer.Lister(),
+		clusterLister:     clusterLister,
+		rbacClient:        rbacClient,
+	}
+
+	return factory.New().
+		WithInformersQueueKeyFunc(func(obj runtime.Object) string {
+			key, _ := cache.MetaNamespaceKeyFunc(obj)
+			return key
+		}, roleInformer.Informer(), roleBindingInformer.Informer()).
+		WithSync(controller.sync).ToController("AddFinalizerController", recorder)
+}
+
+// sync adds finalizer to role/rolebinding for work agent in the spoke cluster namespace. No matter which
+// resource trigger the sync, both role and rolebinding will be reconciled.
+func (m *AddFinalizerController) sync(ctx context.Context, controllerContext factory.SyncContext) error {
+	key := controllerContext.QueueKey()
+
+	// ignore role/rolebinding whose name does not end with the expected suffix
+	if !strings.HasSuffix(key, keySuffix) {
+		return nil
+	}
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		// ignore role/rolebinding whose key is not in format: namespace/name
+		return nil
+	}
+	clusterName := strings.TrimSuffix(name, keySuffix)
+
+	// ignore role/rolebinding for a nonexistent spoke cluster
+	_, err = m.clusterLister.Get(clusterName)
+	if errors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	// ignore role/rolebinding which is not in a cluster namespace
+	if namespace != clusterName {
+		return nil
+	}
+
+	errs := []error{}
+
+	// sync role
+	role, err := m.roleLister.Roles(namespace).Get(name)
+	switch {
+	case errors.IsNotFound(err):
+		// no work if role is not found
+	case err != nil:
+		errs = append(errs, err)
+	default:
+		if err = m.syncRole(ctx, role); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	// sync rolebinding
+	rolebinding, err := m.roleBindingLister.RoleBindings(namespace).Get(name)
+	switch {
+	case errors.IsNotFound(err):
+		// no work if rolebinding is not found
+	case err != nil:
+		errs = append(errs, err)
+	default:
+		if err = m.syncRoleBinding(ctx, rolebinding); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	// aggregate errors if necessary
+	if len(errs) > 0 {
+		err = utilerrors.NewAggregate(errs)
+		klog.Errorf("Reconcile role/rolebinding %s fails with err: %v", key, err)
+		return err
+	}
+	return nil
+}
+
+// syncRole adds finalizer on role for work agent in the spoke cluster namespace
+func (m *AddFinalizerController) syncRole(ctx context.Context, role *rbacv1.Role) error {
+	klog.V(4).Infof("Reconciling Role %s/%s", role.Namespace, role.Name)
+	role = role.DeepCopy()
+
+	// don't add finalizers to instances that are deleted
+	if !role.DeletionTimestamp.IsZero() {
+		return nil
+	}
+
+	// don't add finalizer to instances that already have it
+	for i := range role.Finalizers {
+		if role.Finalizers[i] == manifestWorkFinalizer {
+			return nil
+		}
+	}
+
+	// if this conflicts, we'll simply try again later
```

Comment:
nit: it's best to avoid using words like 'simply', they are subjective. if you just drop 'simply' the comment will be informative.